network/shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java

```diff
@@ -138,6 +139,26 @@ public void registerWithShuffleServer(
     client.sendRpcSync(registerMessage, 5000 /* timeoutMs */);
   }
 
+  /**
+   * Removes this application from the external shuffle server and optionally deletes local
+   * files.
+   *
+   * @param host Host of the shuffle server.
+   * @param port Port of the shuffle server.
+   * @param cleanupLocalDirs True if corresponding shuffle files should be deleted
+   * @throws IOException
+   */
+  public void applicationRemoved(String host,
+      int port,
```

Comment:
The `ExternalShuffleBlockResolver` is the server part. The client part that you see here is actually calling that method, through the RPC layer. This class (the client) is used by executors to connect to other executors for fetching blocks, and offers a synchronous API on top of the network. In **this** case, this client is also used by the driver to communicate to *all* executors that may have shuffle files that need to be deleted.